### PR TITLE
chore: Upgrade Cypress from v9 to v12

### PR DESCRIPTION
**Pull request by LekoArts**
## Description

This updates our E2E tests from Cypress v9 to v12. Migration guide: https://docs.cypress.io/guides/references/migration-guide

I did also do some other following changes while at it:

- Upgrade `start-server-and-test` to v2 (no breaking changes for us)
- Switch from `cypress-image-snapshot` to `@simonsmith/cypress-image-snapshot` because the former isn't compatible with Cypress v12
- Move Cypress configs to TS
- Move support/plugins to TS
- Move Cypress Cloud record keys to env vars
- Move Cypress Cloud project IDs to config files
- Add group names to trailing slash test suite

### Tests

All tests should still run fully and pass.

### CODE REVIEW

**Review by pieh — DISMISSED** [review]

File: e2e-tests/contentful/cypress/integration/gatsby-image-cdn.js [hunk 1]
```diff
@@ -1,5 +1,3 @@
-import { parseSrcset } from "srcset"
```
Comment (LekoArts) [inline e2e-tests/contentful/cypress/integration/gatsby-image-cdn.js:1]: Unused import

File: e2e-tests/contentful/package.json [hunk 2]
```diff
@@ -19,13 +19,15 @@
   },
   "devDependencies": {
     "@cypress/snapshot": "^2.1.7",
+    "@simonsmith/cypress-image-snapshot": "^6.0.1",
```
Comment (LekoArts) [inline e2e-tests/contentful/package.json:22]: cypress-image-snapshot doesn't work with Cypress 12 and this is a fork

File: e2e-tests/contentful/package.json [hunk 3]
```diff
@@ -39,8 +41,8 @@
     "test": "cross-env CYPRESS_SUPPORT=y npm run build && npm run start-server-and-test",
     "start-server-and-test": "start-server-and-test serve http://localhost:9000 cy:run",
     "serve": "gatsby serve",
-    "cy:open": "cypress open",
-    "cy:run": "node ../../scripts/cypress-run-with-conditional-record-flag.js --browser chrome"
+    "cy:open": "cypress open --browser chrome --e2e",
```
Comment (LekoArts) [inline e2e-tests/contentful/package.json:44]: Cypress now also has component testing so this tells it to directly open the e2e test suite with the chrome browser

File: e2e-tests/development-runtime/cypress/integration/styling/less.js [hunk 4]
```diff
@@ -24,12 +24,11 @@ describe(`styling: less`, () => {
   })
 
   describe(`hot reloading`, () => {
-    it(`plain less`, () => {
-      // we don't want to visit page for each test - we want to visit once and then test HMR
-      cy.window().then(win => {
-        cy.spy(win.console, `log`).as(`hmrConsoleLog`)
```
Comment (LekoArts) [inline e2e-tests/development-runtime/cypress/integration/styling/less.js:30]: This is already getting set inside the `support/e2e.ts` file with the

```
Cypress.Commands.overwrite("visit")
```

overwrite

File: e2e-tests/production-runtime/cypress/integration/compilation-hash.js [hunk 5]
```diff
@@ -70,21 +70,20 @@ describe(
...
 
-        // cy.waitForAPI()
-        cy.window().then(window => {
-          // just setting some property on the window
-          // we will later assert that property to know wether
-          // browser reload happened or not.
-          window.notReloaded = true
-          window.___navigate(`/deep-link-page/`)
-        })
+        // just setting some property on the window
```
Comment (LekoArts) [inline e2e-tests/production-runtime/cypress/integration/compilation-hash.js:73]: Adjusted from https://glebbahmutov.com/blog/detect-page-reload/

File: e2e-tests/production-runtime/cypress/integration/lifecycle-methods.js [hunk 6]
```diff
@@ -15,7 +15,7 @@ describe(`Production build tests`, () => {
 
     // we expect 2 `componentDidMount` calls - 1 for initial page and 1 for second page
     cy.lifecycleCallCount(`componentDidMount`).should(`equal`, 2)
-    cy.lifecycleCallCount(`render`).should(`equal`, Cypress.env(`TEST_PLUGIN_OFFLINE`) ? 3 : 2)
+    cy.lifecycleCallCount(`render`).should(`equal`, 2)
```
Comment (LekoArts) [inline e2e-tests/production-runtime/cypress/integration/lifecycle-methods.js:18]: This now works 🤷

File: e2e-tests/contentful/cypress/snapshots/cypress/integration/gatsby-image-cdn.js/gatsby-image-cdn-0.snap.png
Comment (LekoArts) [inline e2e-tests/contentful/cypress/snapshots/cypress/integration/gatsby-image-cdn.js/gatsby-image-cdn-0.snap.png:1]: The paths for the snapshots changed. You can check the E2E test to compare the paths it generates now to this

File: scripts/cypress-run-with-conditional-record-flag.js [hunk 7]
```diff
@@ -10,8 +10,7 @@ const IS_CI = !!(
   false
 )
 
-const shouldRecord =
-  !!process.env.CYPRESS_PROJECT_ID && !!process.env.CYPRESS_RECORD_KEY && IS_CI
+const shouldRecord = !!process.env.CYPRESS_RECORD_KEY && IS_CI
```
Comment (LekoArts) [inline scripts/cypress-run-with-conditional-record-flag.js:13]: projectId is now mostly in the config file itself, so this had to go

File: scripts/cypress-run-with-conditional-record-flag.js [hunk 8]
```diff
@@ -22,6 +21,10 @@ const cypressArgs = [`run`, ...process.argv.slice(2)]
 
 if (shouldRecord) {
   cypressArgs.push(`--record`)
+
+  if (process.env.CYPRESS_GROUP_NAME) {
+    cypressArgs.push(`--group`, process.env.CYPRESS_GROUP_NAME)
```
Comment (LekoArts) [inline scripts/cypress-run-with-conditional-record-flag.js:26]: `--group` fails if `--record` is not also given

**Review by pieh — APPROVED** [review]

**Pull request merged**